### ISSUE TITLE
Polish javadocs for OptionSet and Option classes

### ISSUE THREAD

**Issue by jvalkeal** [eclipse/californium#450]
As agreed in #440 javadocs for particular classes in question should get polished to reflect more closely how specification defines use of options as framework works relatively close on particular details what gets exposed to a user.
Creating this issue for PR to come.

**Comment by jvalkeal** [comment]
@boaks I started to update javadocs for these classes and rfc raises one question.

[3.2. Option Value Formats](https://tools.ietf.org/html/rfc7252#section-3.2) defines types as `empty, opaque, uint, string`. [5.10. Option Definitions](https://tools.ietf.org/html/rfc7252#section-5.10) then gives a list of types mapped to particular options. There is one option `5/If-None-Match` which defines format as `empty` which we don't have enum in `org.eclipse.californium.core.coap.OptionNumberRegistry.optionFormats` which only defines `INTEGER, STRING, OPAQUE, UNKNOWN`.

When looking `org.eclipse.californium.core.coap.OptionNumberRegistry.getFormatByNr(int)`, it returns `IF_NONE_MATCH` as `optionFormats.INTEGER` which I believe is wrong if being academic as that option never have any other value than empty `byte[]`. `Content-Format` and `Accept` value is `0` when empty `byte[]` is given but with `If-None-Match` there really is no real value representation in `org.eclipse.californium.core.coap.Option` methods `getIntegerValue()` or `getStringValue()`. If being precise `getStringValue()` should return `null` and there's nothing we can do about `getIntegerValue()` as its return type is `int` so could not return `null` from that method.

Should we just mention that `5/If-None-Match` is a special type of an option which presense should really be cheched using `org.eclipse.californium.core.coap.OptionSet.hasIfNoneMatch()` as that is more closely what spec defines? Maybe there is a reasoning why `optionFormats` don't have enum for `EMPTY`?

**Comment by vikram919** [comment]
Hi @jvalkeal !
Thanks for your Inverstigation.
you are right, enum optionFormats at `org.eclispe.californium.core.coap.OptionNumberRegistery` should provide empty option value format for If-None-Match option and not Integer option value format as per Spec. 
@boaks Do you agree with it ? or is there any reason for not using empty option value format? 
@jvalkeal  from your comment 1) `Content-Format and Accept value is 0 when empty byte is given` ,  while representing Request/Response in  conventional byte array  not just empty byte but also the length of option is 2 bytes as per Spec.
2) `but with If-None-Match there really is no real value representation in org.eclipse.californium.core.coap.Option methods getIntegerValue() or getStringValue(). If being precise getStringValue() should return null and there's nothing we can do about getIntegerValue() as its return type is int so could not return null from that method` 
for a byte[] empty = new Byte[0]; can take an int value 0; 1 byte can take int values from 0-255.
I fell we need to change the enum for `If-None-Match` option am still working on it to investigate further with some unit tests. correct me if I got it wrong?

**Comment by vikram919** [comment]
Hi @jvalkeal 
from my unit tests 
here are my results,
As per RFC 7252 Section 5.4.2 

> Unrecognized options of class "critical" that occur in a
      Confirmable request MUST cause the return of a 4.02 (Bad Option)
      response

The unrecognised criticial option:` is using If-None-Match option Repeatable `
Bad option response is not returned as response by Dataparser level or on CoapEndpoint level.
Now I conclude it by saying not just the java docs of the option set and option classes have to be polished but we need to fix the If-None-Match `Unrecognized Critical option` in a Confirmable message as a Bad option Diagonistic payload Response.

**Comment by boaks** [comment]
@jvalkeal 

> There is one option 5/If-None-Match which defines format as empty which we don't have enum in 

Seems that "empty" is "newly" introduced (in draft-ietf-core-coap-10, june 2012  :-) ). So I guess, those who implement it, where faced a lot of changes and give up to be up-to-date but also missed it to finalize with the finalization of the RFC7252. 

> Should we ...
 
Right now I would just add  a warning, that EMPTY is not directly supported and if read may result in invalid or unexpected values. Otherwise the improvement of the javadoc would be blocked by that and I would prefer to have that and fix the EMPTY later.

**Comment by boaks** [comment]
@vikram919 

Can you create a new issue about the repeatable options?
I would prefer to isolate this new issue from this issue, otherwise the discussion would explode :-).

**Comment by jvalkeal** [comment]
ok thanks guys, this clarifies what I want to write to javadocs.

**Comment by boaks** [comment]
solved with #461
